### PR TITLE
Hide serf and raft from broker

### LINKED ISSUE #22: Hide Serf and Raft from Broker
Issue: Broker struct is over-burdened.
Proposal: Hide details of serf and raft from Broker behind interfaces
Potential benefits?:
1. Responsibility of broker package becomes leaner, making it easier to navigate
2. Easier to write fine-grained tests of functionalities by mocking out serf and raft interfaces, and separate testing for serf and raft functionalities

Code changes proposal: [Feature branch](https://github.com/travisjeffery/jocko/compare/master...shwetabhgarg:broker_cleanup)
This change set moves serf/raft code from broker to outside. All tests pass.
Next steps (todo):
- write tests by mocking out serf and raft

### PR DESCRIPTION
**Pull request by shwetabhgarg**
Hide serf and raft from broker ( closes #22 )
change set:
- move serf and raft implementations to own packages
- hide serf and raft from broker behind interfaces

next steps (todo):
- independently test serf and raft
- fine-grained tests for jocko functionalities by mocking out serf and raft

### CODE REVIEW

File: broker/fsm.go [hunk 1]
```diff
@@ -1,9 +1,12 @@
 package broker
 
 import (
+	"encoding/json"
 	"io"
 
 	"github.com/hashicorp/raft"
```
Comment (travisjeffery) [inline broker/fsm.go]: it'd be nice if we could make it so the broker pkg didn't depend on raft (and same for serf). we'd have to do some wrapping for things like *raft.Log for the apply method though.

Comment (shwetabhgarg) [inline broker/fsm.go]: I had thought of doing that but this change set was already big for a PR, so I was planning to come back later for it. I'll take your suggestion and make the change now. I'll put a wrapper around serf.MemberStatus and raft.fsm. I'll update here when done.

**Comment by travisjeffery** [comment]
OK just merged most of this but had to fix some conflicts from some other changes from before, so before you work on anything else make a new branch off master so you don't have to deal with conflicts.

**Comment by shwetabhgarg** [comment]
Cool. I'm working on the discussed changes for wrapper around serf.MemberStatus and raft.FSM. I'll update my branch and make a new PR.

**Pull request closed without merging**